cpp/code/basic_arrow.cc

```diff
@@ -70,48 +76,48 @@ TEST(BasicArrow, ReturnNotOk) { ASSERT_OK(ReturnNotOk()); }
 /// Only supports floating point and integral types. Does not support decimals.
 class TableSummation {
   double partial = 0.0;
- public:
 
+ public:
   arrow::Result<double> Compute(std::shared_ptr<arrow::RecordBatch> batch) {
-    for (std::shared_ptr<arrow::Array> array : batch->columns()) {
+    for (const auto& array : batch->columns()) {
       ARROW_RETURN_NOT_OK(arrow::VisitArrayInline(*array, this));
     }
     return partial;
   }
 
   // Default implementation
-  arrow::Status Visit(const arrow::Array& array) {
-    return arrow::Status::NotImplemented("Can not compute sum for array of type ",
-                                         array.type()->ToString());
-  }
-
   template <typename ArrayType, typename T = typename ArrayType::TypeClass>
-  arrow::enable_if_number<T, arrow::Status> Visit(const ArrayType& array) {
-    for (arrow::util::optional<typename T::c_type> value : array) {
-      if (value.has_value()) {
-        partial += static_cast<double>(value.value());
+  arrow::Status Visit(const ArrayType& array) {
+    if constexpr (arrow::is_number_type<T>::value) {
+      for (auto value : array) {
+        if (value.has_value()) {
+          partial += static_cast<double>(value.value());
+        }
       }
+      return arrow::Status::OK();
+    } else {
+      return arrow::Status::NotImplemented("Can not compute sum for array of type ",
+                                           array.type()->ToString());
     }
-    return arrow::Status::OK();
   }
 };  // TableSummation
 
 arrow::Status VisitorSummationExample() {
   StartRecipe("VisitorSummationExample");
-  std::shared_ptr<arrow::Schema> schema = arrow::schema({
```

Comment:
Note that in general, it was consciously chosen to write out return types to be explicit for readers who may be unfamiliar with Arrow. Some things like lambdas and initializing a builder on the stack make sense to simplify to me, but helpers like `schema` should be written out.

Reply:
I see. I've reverted most `auto`s in f069a5be4d6544e26edd6735823e7d0442c87564